### PR TITLE
Create Apple Silicon arm64 wheels

### PR DESCRIPTION
**Pull request by matham**
This adds support for making universal binaries for HDF5 and its deps that supports both arm64 and x86_64 arch in one file. Then, we can use them to make a universal wheel.

Changes:

* Consolidated the manylinux, aarch64 linux, and mac x86/arm64 builds in one action.
* Added caching for the built dependencies so that it'll be significantly faster on aarch64, but also on mac.
* Bumped the HDF5 version back to 1.12.1.
* Compiling mac arm64 and x86_64 binaries, including hdf5 and its dependencies as individual wheels for each platform.
* To support arm64 cross-compiling I had to apply the patches from [here](https://github.com/conda-forge/hdf5-feedstock/commit/2cb83b63965985fa8795b0a13150bf0fd2525ebd) (thanks conda-forge). Neither 1.13 nor 1.12 support cross-compiling just arm64. And I suspect that that fat binaries I build previously didn't work at all on arm64, even though I could build it.

### CODE REVIEW

**Comment by matham** [comment]
I tested the updated build script by running each of the components and it worked. I did not build a wheel though, but I i think the wheel part should just work. You can see how the generated libs ineed contain both arm64 and x86_64 components:


```
$ file /tmp/hdf5/lib/*
/tmp/hdf5/lib/cmake:                directory
/tmp/hdf5/lib/libbenchmark.a:       Mach-O universal binary with 2 architectures: [x86_64:current ar archive] [arm64]
/tmp/hdf5/lib/libbenchmark.a (for architecture x86_64): current ar archive
/tmp/hdf5/lib/libbenchmark.a (for architecture arm64):  current ar archive
/tmp/hdf5/lib/libbenchmark_main.a:  Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libbenchmark_main.a (for architecture x86_64):    current ar archive random library
/tmp/hdf5/lib/libbenchmark_main.a (for architecture arm64):     current ar archive random library
/tmp/hdf5/lib/libbz2.a:             current ar archive random library
/tmp/hdf5/lib/libgmock.a:           Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libgmock.a (for architecture x86_64):     current ar archive random library
/tmp/hdf5/lib/libgmock.a (for architecture arm64):      current ar archive random library
/tmp/hdf5/lib/libgmock_main.a:      Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libgmock_main.a (for architecture x86_64):        current ar archive random library
/tmp/hdf5/lib/libgmock_main.a (for architecture arm64): current ar archive random library
/tmp/hdf5/lib/libgtest.a:           Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64]
/tmp/hdf5/lib/libgtest.a (for architecture x86_64):     current ar archive random library
/tmp/hdf5/lib/libgtest.a (for architecture arm64):      current ar archive random library
/tmp/hdf5/lib/libgtest_main.a:      Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libgtest_main.a (for architecture x86_64):        current ar archive random library
/tmp/hdf5/lib/libgtest_main.a (for architecture arm64): current ar archive random library
/tmp/hdf5/lib/libhdf5.300.dylib:    Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64]
/tmp/hdf5/lib/libhdf5.300.dylib (for architecture x86_64):      Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libhdf5.300.dylib (for architecture arm64):       Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libhdf5.a:            Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64]
/tmp/hdf5/lib/libhdf5.a (for architecture x86_64):      current ar archive random library
/tmp/hdf5/lib/libhdf5.a (for architecture arm64):       current ar archive random library
/tmp/hdf5/lib/libhdf5.dylib:        Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64]
/tmp/hdf5/lib/libhdf5.dylib (for architecture x86_64):  Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libhdf5.dylib (for architecture arm64):   Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libhdf5.la:           libtool library file, ASCII text
/tmp/hdf5/lib/libhdf5.settings:     ASCII text, with very long lines
/tmp/hdf5/lib/libhdf5_hl.300.dylib: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libhdf5_hl.300.dylib (for architecture x86_64):   Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libhdf5_hl.300.dylib (for architecture arm64):    Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libhdf5_hl.a:         Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libhdf5_hl.a (for architecture x86_64):   current ar archive random library
/tmp/hdf5/lib/libhdf5_hl.a (for architecture arm64):    current ar archive random library
/tmp/hdf5/lib/libhdf5_hl.dylib:     Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libhdf5_hl.dylib (for architecture x86_64):       Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libhdf5_hl.dylib (for architecture arm64):        Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libhdf5_hl.la:        libtool library file, ASCII text
/tmp/hdf5/lib/liblz4.1.9.3.dylib:   Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/liblz4.1.9.3.dylib (for architecture x86_64):     Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/liblz4.1.9.3.dylib (for architecture arm64):      Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/liblz4.1.dylib:       Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/liblz4.1.dylib (for architecture x86_64): Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/liblz4.1.dylib (for architecture arm64):  Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/liblz4.a:             Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/liblz4.a (for architecture x86_64):       current ar archive random library
/tmp/hdf5/lib/liblz4.a (for architecture arm64):        current ar archive random library
/tmp/hdf5/lib/liblz4.dylib:         Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/liblz4.dylib (for architecture x86_64):   Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/liblz4.dylib (for architecture arm64):    Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/liblzo2.a:            Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/liblzo2.a (for architecture x86_64):      current ar archive random library
/tmp/hdf5/lib/liblzo2.a (for architecture arm64):       current ar archive random library
/tmp/hdf5/lib/libsnappy.a:          Mach-O universal binary with 2 architectures: [x86_64:current ar archive] [arm64:current ar archive]
/tmp/hdf5/lib/libsnappy.a (for architecture x86_64):    current ar archive
/tmp/hdf5/lib/libsnappy.a (for architecture arm64):     current ar archive
/tmp/hdf5/lib/libz.1.2.11.dylib:    Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libz.1.2.11.dylib (for architecture x86_64):      Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libz.1.2.11.dylib (for architecture arm64):       Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libz.1.dylib:         Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libz.1.dylib (for architecture x86_64):   Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libz.1.dylib (for architecture arm64):    Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libz.a:               Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libz.a (for architecture x86_64): current ar archive random library
/tmp/hdf5/lib/libz.a (for architecture arm64):  current ar archive random library
/tmp/hdf5/lib/libz.dylib:           Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libz.dylib (for architecture x86_64):     Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libz.dylib (for architecture arm64):      Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libzstd.1.5.2.dylib:  Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libzstd.1.5.2.dylib (for architecture x86_64):    Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libzstd.1.5.2.dylib (for architecture arm64):     Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libzstd.1.dylib:      Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libzstd.1.dylib (for architecture x86_64):        Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libzstd.1.dylib (for architecture arm64): Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/libzstd.a:            Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
/tmp/hdf5/lib/libzstd.a (for architecture x86_64):      current ar archive random library
/tmp/hdf5/lib/libzstd.a (for architecture arm64):       current ar archive random library
/tmp/hdf5/lib/libzstd.dylib:        Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/hdf5/lib/libzstd.dylib (for architecture x86_64):  Mach-O 64-bit dynamically linked shared library x86_64
/tmp/hdf5/lib/libzstd.dylib (for architecture arm64):   Mach-O 64-bit dynamically linked shared library arm64
/tmp/hdf5/lib/pkgconfig:            directory
```

**Comment by avalentino** [comment]
> Hmm I think the ci/** branch builds wheels only for branches in this repo, not forks..

you have the wheels built in your fork

**Comment by matham** [comment]
Yeah, I just earlier realized that I can get it in the fork actions, but I had to enable it first. It seems to be progressing well! I'll post when I have wheels that can be tested.

**Comment by matham** [comment]
Phew finally done!

I updated the first comment with the current changes. You can see the succeeded builds here: https://github.com/matham/PyTables/actions/runs/1973006597.

I'm not on a mac so would appreciate if someone can actually try it to see if works on a real machine, in particular on the M1, including the compression dependencies. You can get the wheels from the action (see artifact at the bottom).

When merged, a build from a wheel generated from the cache should also be tested to be sure it's not missing anything.

**Comment by avalentino** [comment]
Thanks @matham.
@xmatthias, @amotl I would appreciate your review of this PR.

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -153,24 +114,30 @@ jobs:
...
+          CIBW_ENVIRONMENT_MACOS: CC=/usr/bin/clang CXX=/usr/bin/clang LZO_DIR=/tmp/hdf5 BZIP2_DIR=/tmp/hdf5 HDF5_DIR=/tmp/hdf5 HDF5_VERSION=${{ env.HDF5_VERSION }} CFLAGS=-g0 LD_LIBRARY_PATH="/tmp/hdf5/lib:${LD_LIBRARY_PATH}" PKG_CONFIG_PATH="/tmp/hdf5/lib/pkgconfig:${PKG_CONFIG_PATH}"
+          CIBW_SKIP: "*-musllinux_* cp38-macosx_x86_64 cp39-macosx_x86_64 cp310-macosx_x86_64"
+          CIBW_BEFORE_ALL_MACOS: cp -r `pwd`/hdf5_build /tmp/hdf5
+          CIBW_BEFORE_ALL_LINUX: >
+            cp -r `pwd`/hdf5_build /tmp/hdf5 &&
+            yum -y update &&
+            yum install -y zlib-devel bzip2-devel lzo-devel
+          CIBW_REPAIR_WHEEL_COMMAND_MACOS: >
+            DYLD_FALLBACK_LIBRARY_PATH=/tmp/hdf5/lib delocate-listdeps {wheel} &&
+            DYLD_FALLBACK_LIBRARY_PATH=/tmp/hdf5/lib delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}
```
Comment (xmatthias) [inline .github/workflows/wheels.yml:133]: you seem to set `DYLD_FALLBACK_LIBRARY_PATH` twice - so assuming it works, i guess you can remove the first part?

File: ci/github/get_hdf5.sh [hunk 2]
```diff
@@ -0,0 +1,135 @@
+#!/bin/bash
+# vendored from https://github.com/h5py/h5py/blob/master/ci/get_hdf5_if_needed.sh
```
Comment (xmatthias) [inline ci/github/get_hdf5.sh:2]: This comment does't seem to be true anymore (it seems this file experienced quite some changes).

That's not per se bad - but maybe we can change the wording to "inspired by ..." ?

File: ci/github/get_hdf5.sh [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    curl -sLO https://github.com/lz4/lz4/archive/refs/tags/v1.9.3.tar.gz
+    tar xzf v1.9.3.tar.gz
+    pushd lz4-1.9.3
+    make install PREFIX="$HDF5_DIR"
+    popd
+
+    # bzip2
+    curl -sLO https://gitlab.com/bzip2/bzip2/-/archive/bzip2-1.0.8/bzip2-bzip2-1.0.8.tar.gz
+    tar xzf bzip2-bzip2-1.0.8.tar.gz
+    pushd bzip2-bzip2-1.0.8
```
Comment (xmatthias) [inline ci/github/get_hdf5.sh:82]: You might want to use a variable for the different versions (not just in the bzip2 case) to make future updates easier.

if i were to update this to 1.0.9 - i'd be at least 7 lines i'd need to change - so 7 possibilities for either a typo or a "forgot to change".

**Review by xmatthias — COMMENTED** [review]
aside from the nit-picking in the comments above/below

I can confirm it's working on linux x64 and aarch64 - and the changes themselves seem to make sense (and remove the dependency on brew - which is good).

Nothing in the windows build changed - so i guess that's fine.
I'm unable to test anything on macOS.

Maybe a look from @cgohlke would make sense, too.
his experience in the python build system (when it comes to integrating binary libraries into wheels) is unbeaten - at least by people i know about.

**Comment by avalentino** [comment]
Thanks @xmatthias

File: .github/workflows/wheels.yml (hunk 1 above)
Comment (matham) [reply .github/workflows/wheels.yml:133]: I'm not quite sure what you mean? I think `DYLD_FALLBACK_LIBRARY_PATH` is needed for both `delocate-listdeps` and `delocate-wheel` commands.

File: ci/github/get_hdf5.sh (hunk 2 above)
Comment (matham) [reply ci/github/get_hdf5.sh]: Fixed

File: ci/github/get_hdf5.sh (hunk 3 above)
Comment (matham) [reply ci/github/get_hdf5.sh]: Fixed

**Comment by matham** [comment]
Here's the latest build incorporating the feedback changes: https://github.com/matham/PyTables/actions/runs/1977575823

File: .github/workflows/wheels.yml (hunk 1 above)
Comment (xmatthias) [reply .github/workflows/wheels.yml:133]: you're right, my bad (i overlooked the space there)

**Comment by matham** [comment]
[Apparently](https://github.com/h5py/h5py/pull/2065) HDF5-1.13 is an experimental release and is described here: https://www.hdfgroup.org/2021/12/hdf5-1-13-0-introducing-experimental-releases/.

I'm not sure if you want to revert to 1.12 and wait for 1.14 to add universal support or continue on.

**Comment by avalentino** [comment]
Not sure. Maybe we could have a new workflow for experimental builds and still keep the old one for stable builds.

**Comment by xmatthias** [comment]
> Our next maintenance release, HDF5 1.14.0, starts new series of maintenance releases. It is targeted for release in late 2022

While i'm not personally affected - having to wait for "almost a year" in case of problems is quite troublesome in many instances - and usually the usecase will be gone by then.

> Not sure. Maybe we could have a new workflow for experimental builds and still keep the old one for stable builds.

That's a possibility - but i'm not sure about the benefits we expect from that.

While the test-suite will run - in my experience a test-suite will almost never capture all edge-cases - so that won't guarantee that pytables is automatically compatible with the new release.

As it's experimental, you couldn't/shouldn't use it for stable releases either.

alternatively pytables might start to do experimental / pre-releases for people with problems on macos - so these could install with `pip install tables==3.8.0.dev0` [source: PEP440](https://peps.python.org/pep-0440/#developmental-releases) as a workaround until this propagates to the stable branch.

That'd however require additional releases -> additional time resources - which i'm not sure are available.

**Comment by matham** [comment]
Yeah, given that the file itself could be broken in a future release if the internals change, which they say it might for experimental releases, even end users probably won't benefit from 1.13 as any files they create may not work in 1.14.

So I guess it's best to wait for 1.14 for this, sadly.

**Comment by matham** [comment]
I could potentially still make the universal binaries for the deps, except for hdf5 itself and build separate x86/arm64 wheels instead of a universal wheel. That I imagine still works for 1.12 (I hope).

Then, once 1.14 comes out we can just change the cibw env variable to build the universal wheels instead of the separate. BTW, the downside of not having universal wheels, in case it wasn't clear, is that with universal wheels as long as they exist, I can download the wheels on github x86 CI and packages them there and it'll work on arm64. But when they are separate you need an actual arm64 to package for arm64.

**Comment by matham** [comment]
Ok, after some further investigation I have given up on the idea of universal wheels and instead focused just on building arm64 and x86 wheels for the python versions that support them. Ultimately to make universal2 wheels, `delocate` now has a fuse option that can just merge the two wheels, but packaging libraries can add that functionality and it doesn't have to be added here.

To be able to cross compile on x86 for arm64 I had to apply [these patches](https://github.com/conda-forge/hdf5-feedstock/commit/2cb83b63965985fa8795b0a13150bf0fd2525ebd) from conda-forge. Neither 1.12  not 1.13 would otherwise cross-compile just arm64. I suspect my previous universal wheels, even though it built would probably never have run on apple silicon.

Are these patches acceptable? Currently I download them and apply them from the original commit, but I could add them to the repo if that is desired. I'll bring this up on [HDF5's repo](https://github.com/HDFGroup/hdf5/issues/566), but I'm not sure whether they are likely to fix cross-compiling (hopefully). Otherwise, to support it, for newer versions of HDF5, we'd probably need to upgrade the patches (or just wait until conda-forge does it 😁).

I tested the wheels from [this build](https://github.com/matham/PyTables/actions/runs/2002048525) on a Scaleaway M1 I spun up:
```py
Python 3.9.10 (main, Jan 15 2022, 11:40:53)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tables
>>> from tables import *
>>> h5file = open_file("tutorial1.h5", mode="w", title="Test file")
>>> group = h5file.create_group("/", 'detector', 'Detector information')
>>> class Particle(IsDescription):
...     name      = StringCol(16)
...
>>> table = h5file.create_table(group, 'readout', Particle, "Readout example")
>>> print(h5file)
tutorial1.h5 (File) 'Test file'
Last modif.: '2022-03-18T05:01:47+00:00'
Object Tree:
/ (RootGroup) 'Test file'
/detector (Group) 'Detector information'
/detector/readout (Table(0,)) 'Readout example'

>>> h5file
File(filename=tutorial1.h5, title='Test file', mode='w', root_uep='/', filters=Filters(complevel=0, shuffle=False, bitshuffle=False, fletcher32=False, least_significant_digit=None))
/ (RootGroup) 'Test file'
/detector (Group) 'Detector information'
/detector/readout (Table(0,)) 'Readout example'
  description := {
  "name": StringCol(itemsize=16, shape=(), dflt=b'', pos=0)}
  byteorder := 'little'
  chunkshape := (4096,)

>>> import platform
>>> platform.uname()
uname_result(system='Darwin', node='...', release='21.1.0', version='Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:24 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T8101', machine='arm64')
```

I feel like if the patches are acceptable then this should be done. The most recent built wheels (still finishing) can be downloaded from [here](https://github.com/matham/PyTables/actions/runs/2002216391).

**Review by avalentino — APPROVED** [review]

**Comment by avalentino** [comment]
Thanks @matham
